### DISCUSSION TITLE
How to expose colima env via tcp?

### DISCUSSION THREAD

**Discussion by mikhail-gaponov** [abiosoft/colima#361] [Q&A]
Hi Colima Community!
   I've tried to find out a way to expose colima profile runtime via TCP and I didn't find it at colima docs or Internet.
Could somebody show me the right direction to move forward?

As for the final goal I'd like to have separate local docker environments.

**Comment by abiosoft** [comment]
What do you mean by expose via TCP? 
If you referring to a reachable IP address, then that can be enabled via the `--network-address` flag.

```
colima start --network-address
```

You can view the IP address for each profile with

```
colima list
```

**Reply by mikhail-gaponov** [reply]
Hi @abiosoft! 
Thanks for the answer! The "--network-address" option assigns/exposes VM's IP outside but it's not the goal I'm trying to achieve in scope of this question.
My question is about docker client interaction with docker-complient engine. Each colima-created VM maps to particular docker context and I can see all contexts via "docker context ls" command.
So, in order to use desired VM, I have to switch context via docker command:
    % _docker context use <context name>_
    % _docker <command for active context>_
or 
    % _docker -H <unix:///entry point> <command>_

   Colima creates unix socket as a default entry point to communicate with docker client. Docker allows to communicate with docker engine via tcp entrypoint.
   So, my question is how to get access to colima-created docker context via tcp:///IP:PORT (not "unix:///..." entrypoint).

**Reply by abiosoft** [reply]
Unix socket is actually the preferred option as you wouldn't need to worry about port availability and TLS certificates.

Do you want to share the docker instance with another VM or you need to expose it to external network?

**Reply by mikhail-gaponov** [reply]
Well, how deep the rabbit hole is.... I have both needs with next schema:
I have 1 Macbook and from 1 to 3 linux-based bare-metal computers in a local network to buid dev/work infrastructure.
Use cases:
   Case 1: I use a Portainter tool to have a UI for docker environments. It allows to manage the single local docker env via unix socket and multiple ones via tcp.
   Case 2: CI/CD tools (gitlab, jenkins) could (and will) be installed outside of local Macbook machine but with ability to manage containers on it.

Both cases require docker API exposed via tcp.

**Reply by abiosoft** [reply]
Now I understand.

You can modify the docker daemon settings via `colima start --edit` and listen on both socket and port. Kindly check the FAQ

Ensure to listen on the default socket as well, otherwise the local docker context would be broken because Colima makes the assumption.

I am limited to mobile for some days and unable to respond comprehensively.

Thanks.

**Reply by canemacchina** [reply]
what do you mean exactly by "Ensure to listen on the default socket as well"?
I also have the need to connect to the docker daemon through TCP, so I tried this config:
```
docker: {"hosts": ["tcp://0.0.0.0:2375", "unix:///Users/lorenzo/.colima/default/docker.sock"]}
```
But colima fails to start with this error:
```
INFO[0000] starting colima
INFO[0000] runtime: docker
INFO[0000] preparing network ...                         context=vm
INFO[0000] starting ...                                  context=vm
INFO[0021] provisioning ...                              context=docker
INFO[0021] starting ...                                  context=docker
FATA[0083] error starting docker: error running [lima sudo docker info], output: "errors pretty printing info", err: "exit status 1"
```

Instead, everything seems perfectly working using this daemon config
```
docker:  {"hosts": ["tcp://0.0.0.0:2375", "unix:///var/run/docker.sock"]}
```
That doesn't make sense to me because `unix:///var/run/docker.sock` does not exist

**Reply by abiosoft** [reply]
> That doesn't make sense to me because `unix:///var/run/docker.sock` does not exist

@canemacchina 

It exists :) but in the VM, not on your machine. The one on your machine `unix:///Users/lorenzo/.colima/default/docker.sock` is actually being forwarded to `unix:///var/run/docker.sock` in the VM.

The docker config is for the VM, so it needs to point to the socket in the VM.

**Reply by canemacchina** [reply]
ah ok perfect! Thanks!

**Comment by tobias-schuele** [comment]
As described above, I added `docker:  {"hosts": ["tcp://0.0.0.0:2375", "unix:///var/run/docker.sock"]}` to the config. However, when starting Colima with `colima start --network-address`, I get the following error messages (both with 0.6.6 and HEAD):

```
FATA[0272] error starting docker: error running [lima sudo service docker start], output: "Warning: The unit file, source configuration file or drop-ins of docker.service changed on disk. Run 'systemctl daemon-reload' to reload units.", err: "exit status 1"
```

Any ideas what's going wrong? Thanks!

**Reply by stowns** [reply]
same

**Reply by matdan9** [reply]
I had the same problem. Looks like it's related to systemd not reloading the dockerd config. Deleting the colima VM `colima delete` and recreating it `colima start --edit` seems to fix the problem.

**Reply by Crown0815** [reply]
Even after deleting the VM the issue seems to persist:

```sh
$ colima delete
are you sure you want to delete colima and all settings? [y/N] y
this will delete ALL container data. Are you sure you want to continue? [y/N] y
INFO[0002] deleting colima                              
INFO[0002] deleting ...                                  context=docker
INFO[0004] done
$ colima start --edit
INFO[0000] editing in vim                               
INFO[0077] starting colima                              
INFO[0077] runtime: docker                              
INFO[0079] creating and starting ...                     context=vm
INFO[0090] provisioning ...                              context=docker
> Job for docker.service failed because the control process exited with error code.
> See "systemctl status docker.service" and "journalctl -xeu docker.service" for details.
WARN[0091] error restarting docker: exit status 1        context=docker
INFO[0091] starting ...                                  context=docker
FATA[0249] error starting docker: error running [lima sudo service docker start], output: "A dependency job for docker.service failed. See 'journalctl -xe' for details.", err: "exit status 1" 
```

I simply replaced the entry

```
docker: {}
```

with 

```
docker:  {"hosts": ["tcp://0.0.0.0:2375", "unix:///var/run/docker.sock"]}
```

**Reply by Crown0815** [reply]
The problem from journalctl is:

```
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: hosts: (from flag: [fd://], from file: [tcp://0.0.0.0:2375 unix:///var/run/docker.sock])
```

looks like the flag is interfering.

**Reply by Crown0815** [reply]
Indeed the docker service config files contains the problematic entry:

```
$ cat /etc/systemd/system/docker.service.d/docker.conf 

[Service]
LimitNOFILE=infinity
ExecStart=
ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --host-gateway-ip=192.168.5.2
```

(problematic entry: `-H fd://`)

**Reply by Crown0815** [reply]
Removing the `-H fd://` entry from the `/etc/systemd/system/docker.service.d/docker.conf` file did not resolve the issue. After restarting colima, the entry is simply back.

**Reply by jgangemi** [reply]
has anyone found a workaround for this? i see this issue has been created but there has been no movement on it

https://github.com/abiosoft/colima/issues/1174

**Reply by Masqueey** [reply]
This is not working for me, but for a lot of people it is: https://github.com/abiosoft/colima/discussions/565#discussioncomment-7409567
